### ISSUE TITLE
[str4d's standalone miner fork] Invalid shares after connection reset

### ISSUE THREAD

**Issue by daira** [zcash/zcash#2381]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [sammy007](https://github.com/sammy007)**
_Tuesday Nov 01, 2016 at 12:45 UTC_
_Originally opened as https://github.com/str4d/zcash/issues/14_

----

If pool resets connections (restart) miner (and zogminer) will reconnect and all shares will be invalid on pool side.

**Comment by daira** [comment]
<a href="https://github.com/mmitech"><img src="https://avatars0.githubusercontent.com/u/4456920?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mmitech](https://github.com/mmitech)**
_Tuesday Nov 01, 2016 at 14:02 UTC_

----

I can reproduce this as well

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Tuesday Nov 01, 2016 at 14:02 UTC_

----

Related because code is the same https://github.com/nginnever/zogminer/issues/65

**Comment by daira** [comment]
<a href="https://github.com/str4d"><img src="https://avatars3.githubusercontent.com/u/4993799?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [str4d](https://github.com/str4d)**
_Wednesday Nov 02, 2016 at 17:25 UTC_

----

My guess is that after the reconnect, there's an assumption being made either by my Stratum code or the pool about what data is persisted and what is regenerated. Most likely, the cause is that the server provides a new nonce1 to the miner, but the miner is not seeing a new job and so continues mining with the old nonce1. I haven't yet reviewed my code though, and I'd need to see a connection trace to be sure.

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Wednesday Nov 02, 2016 at 17:35 UTC_

----

New connection on pool side, so new nonce is generated and sent on `mining.subscribe` followed by a target and job after `mining.authorize`. Session ID is `null` of course.

**Comment by daira** [comment]
<a href="https://github.com/str4d"><img src="https://avatars3.githubusercontent.com/u/4993799?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [str4d](https://github.com/str4d)**
_Wednesday Nov 02, 2016 at 17:38 UTC_

----

Hmm, then the miner _should_ get notified about the new job and switch to it. I'll look at this later today.

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Wednesday Nov 02, 2016 at 17:56 UTC_

----

This is very necessary to have reference implementation 100% robust and compliant, we are suppering from this mess already, for example zogminer borrowed this stratum code and already ignores `notify` after `set_target` after `authorize` and mining only starts when new job pushed from server (on new block in network).

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Thursday Nov 03, 2016 at 15:11 UTC_

----

I sniffed traffic on port, hope it helps:

```
T 192.168.1.2:9998 -> 192.168.1.4:44040 [AP]
  {"id":null,"method":"mining.notify","params":["699187805179202" ... }
##
T 192.168.1.4:44040 -> 192.168.1.2:9998 [A]
  {"id": 4, "method": "mining.submit", "params": ["X","699187805179202" ... }
##
T 192.168.1.2:9998 -> 192.168.1.4:44040 [AP]
  {"id":4,"result":true,"error":null}.

######## STRATUM RESET ########

T 192.168.1.2:9998 -> 192.168.1.4:44042 [AR]
  ......
####
T 192.168.1.4:44044 -> 192.168.1.2:9998 [AP]
  {"id": 1, "method": "mining.subscribe", "params": ["/MagicBean:1.0.0/", null,"192.168.1.2","9998"]}.
###
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":1,"result":[null,"000000000000000000000000051638b6"],"error":null}.
##
T 192.168.1.4:44044 -> 192.168.1.2:9998 [AP]
  {"id": 2, "method": "mining.authorize", "params": ["X","x"]}.
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":2,"result":true,"error":null}.
#
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":null,"method":"mining.set_target","params":["015555555555555555555555555555555555555555555555555555555555552a"]}.
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":null,"method":"mining.notify","params":["762369330492885","04000000","...","...","00000000000 ... ","7c511b58","4af6031d",true]}.
#
{"id": 4, "method": "mining.submit", "params": ["X","762369330492885","7c511b58" ... }
##
T 192.168.1.2:9998 -> 192.168.1.4:44044 [AP]
  {"id":4,"result":null,"error":[23,"Low difficulty share",null]}.
```

Nonce is new after pool reset, it's randomised for sharding.

**Comment by daira** [comment]
<a href="https://github.com/str4d"><img src="https://avatars3.githubusercontent.com/u/4993799?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [str4d](https://github.com/str4d)**
_Thursday Nov 03, 2016 at 15:30 UTC_

----

Looks like you are using my standalone miner for testing. Could you add `-debug=stratum` to its CLI and then post the submission output? You should be able to see the full nonce (including server portion) when it makes a submission.

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Thursday Nov 03, 2016 at 16:20 UTC_

----

I added `-debug=stratum -printtoconsole`

```
Solution found; Submitting to 192.168.1.2...
  d5555555555555555555555555555561b9381605000000000000000000000000
B-) Submitted and accepted.
[WARN] read_until: End of file
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
[ERROR] Could not connect to stratum server 192.168.1.2:9998, Connection refused
Reconnecting in 3 seconds...
[WARN] read_until: Transport endpoint is not connected
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
Connected!
Subscribed to stratum server
Authorized worker X
Target set to 015555555555555555555555555555555555555555555555555555555555552a
Received new job #371186289322138
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  55555555555555555555555555555594b9381605000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  55555555555555555555555555555594b9381605000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
```

We can see it clearly that old part of nonce is used.

BTW, miner still mine even if server is down, I see that with `-debug=pow`

**Comment by daira** [comment]
<a href="https://github.com/str4d"><img src="https://avatars3.githubusercontent.com/u/4993799?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [str4d](https://github.com/str4d)**
_Thursday Nov 03, 2016 at 17:37 UTC_

----

@sammy007 could you please apply the following patch to my standalone miner and then test it again:

``` diff
diff --git a/src/libstratum/ZcashStratum.cpp b/src/libstratum/ZcashStratum.cpp
index 7315fcc..2042d30 100644
--- a/src/libstratum/ZcashStratum.cpp
+++ b/src/libstratum/ZcashStratum.cpp
@@ -319,12 +319,14 @@ void ZcashMiner::stop()
 void ZcashMiner::setServerNonce(const Array& params)
 {
     auto n1str = params[1].get_str();
+    LogPrintf("New nonce1 received from server: \"%s\"\n", n1str);
     std::vector<unsigned char> nonceData(ParseHex(n1str));
     while (nonceData.size() < 32) {
         nonceData.push_back(0);
     }
     CDataStream ss(nonceData, SER_NETWORK, PROTOCOL_VERSION);
     ss >> nonce1;
+    LogPrintf("Updated nonce template to %s\n", nonce1.GetHex());

     nonce1Size = n1str.size();
     size_t nonce1Bits = nonce1Size * 4; // Hex length to bit length
@@ -387,6 +389,8 @@ ZcashJob* ZcashMiner::parseJob(const Array& params)
     ret->nonce2Space = nonce2Space;
     ret->nonce2Inc = nonce2Inc;

+    LogPrintf("Applied nonce template %s to job %s\n", nonce1.GetHex(), ret->job);
+
     return ret;
 }

```

**Comment by daira** [comment]
<a href="https://github.com/sammy007"><img src="https://avatars3.githubusercontent.com/u/7374093?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [sammy007](https://github.com/sammy007)**
_Thursday Nov 03, 2016 at 17:46 UTC_

----

```
Solution found; Submitting to 192.168.1.2...
  bffffffffffffffffffffffffffffffc4789b601000000000000000000000000
B-) Submitted and accepted.
[WARN] read_until: End of file
Reconnecting in 3 seconds...
Connecting to stratum server 192.168.1.2:9998
Connected!
Subscribed to stratum server
New nonce1 received from server: "000000000000000000000000034c09b9"
Updated nonce template to 00000000000000000000000000000000b9094c03000000000000000000000000
Authorized worker X
Target set to 015555555555555555555555555555555555555555555555555555555555552a
Applied nonce template 00000000000000000000000000000000b9094c03000000000000000000000000 to job 114231506221785
Received new job #114231506221785
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  4000000000000000000000000000000b4789b601000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
Found solution satisfying the server target
Solution found; Submitting to 192.168.1.2...
  800000000000000000000000000000094789b601000000000000000000000000
Low difficulty share
[WARN] :-( Not accepted.
```

tl;dr new nonce **_applied**_ but stale nonce still used.

**Comment by daira** [comment]
<a href="https://github.com/str4d"><img src="https://avatars3.githubusercontent.com/u/4993799?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [str4d](https://github.com/str4d)**
_Thursday Nov 03, 2016 at 17:51 UTC_

----

This is useful, thanks. The confusing thing here is that you're setting `CLEAN_JOBS` to true, which means that the miners should cancel their current work and pick up the new job immediately. But even after a few stales, the new job should be detected. Time to go bug-hunting...

**Comment by nuttycom** [comment]
Closing as stale.
